apps/passport-client/components/screens/ImportBackupScreen.tsx

```diff
@@ -0,0 +1,386 @@
+            if (userHasEmailPCD && pcd.type === EmailPCDTypeName) {
+              return false;
+            }
+
+            // If a PCD with this ID exists already, don't import it
+            if (existingPcdCollection.hasPCDWithId(pcd.id)) {
+              return false;
+            }
+
+            // Otherwise, do import it
+            return true;
+          };
+
+          // These are the PCDs that could be merged, e.g. not duplicates of
+          // existing PCDs, or secondary semaphore identities or emails.
+          const mergeablePcds: PCD[] = parsedCollection
+            .getAll()
+            .filter(preImportFilter);
+
+          // Create a map of the folders these PCDs belong to, to a count of
+          // the number of PCDs in each folder, with `NoFolderSymbol` used for
+          // PCDs belonging to no folder.
+          const pcdFolders: Record<string | symbol, number> =
+            mergeablePcds.reduce((folders, pcd) => {
+              const folder =
+                parsedCollection.getFolderOfPCD(pcd.id) ?? NoFolderSymbol;
+              if (folder in folders) {
+                folders[folder]++;
+              } else {
+                folders[folder] = 1;
+              }
+              return folders;
+            }, {});
+
+          // The set of folders that the user has chosen to import.
+          let selectedFolders: Set<string | symbol>;
+
+          if (importState.state === "initial") {
+            // By default all folders are selected.
+            selectedFolders = new Set([
+              NoFolderSymbol,
+              ...Object.keys(pcdFolders)
+            ]);
+          } else {
+            // Otherwise, make sure previously selected folders are still
+            // valid given current parsed file contents.
+            selectedFolders = importState.selectedFolders;
+            for (const folder of selectedFolders) {
+              // Do we have a selected folder that doesn't exist any more
+              // for some reason?
+              if (!(folder in pcdFolders)) {
+                selectedFolders.delete(folder);
+              }
+            }
+          }
+
+          setImportState({
+            state: "valid-file-selected",
+            collection: parsedCollection,
+            mergeablePcdIds: new Set(mergeablePcds.map((pcd) => pcd.id)),
+            selectedFolders,
+            // Check which of the mergeable PCDs are in selected folders, and
+            // create a set of their IDs
+            selectedPcdIds: new Set(
+              mergeablePcds
+                .filter((pcd) => {
+                  const folder =
+                    parsedCollection.getFolderOfPCD(pcd.id) ?? NoFolderSymbol;
+                  return selectedFolders.has(folder);
+                })
+                .map((pcd) => pcd.id)
+            ),
+            folderCounts: pcdFolders
+          });
+        }
+      }
+    })();
+  }, [filesContent, importState, existingPcdCollection]);
```

Comment:
It feels like a lot of the stuff in this `useEffect` hook may be better suited to live inside of the merge algorithm in `dispatch`, but I also see why that's not the case sometimes.